site/_posts/2020-11-15-how-to-securely-implement-oauth-rails.md

```diff
@@ -0,0 +1,300 @@
+
+{% include _image.liquid src="/assets/img/blogs/fusionauth-example-rails/edit-application.png" alt="The OAuth tab in the Edit Application page of FusionAuth." class="img-fluid" figure=false %}
+
+### Routes
+For this app, we are going to setup three routes:
+* **welcome page**: Our root page or `view`. 
+* **oauth2-callback** endpoint: Receiver of the redirect after logging into FusionAuth.
+* **logout** endpoint: Receiver of the redirect after logging out of FusionAuth.
+
+```ruby
+Rails.application.routes.draw do
+  root to: 'welcome#index'
+  get '/oauth2-callback', to: 'oauth#oauth_callback'
+  get '/logout', to: 'oauth#destroy'
+end
+```
+
+### Authenticating and Logging Out 
+We will start off with a very basic welcome view that contains a link for a user to login to the application.
+
+```ruby
+# app/views/welcome/index.html.erb
+
+<p>Welcome user! Please <%= link_to 'Log In', @login_url %></p>
+```
+
+It should look something like this when rendered.
+{% include _image.liquid src="/assets/img/blogs/fusionauth-example-rails/home.png" alt="The example Home screen for the Rails application." class="img-fluid" figure=false %}
+
+To begin the login process, we will want to construct a URL to make an [Authorization Code Grant Request](https://fusionauth.io/docs/v1/tech/oauth/endpoints/#authorization-code-grant-request).
+This URL is composed of the `client_id`, `client_secret`, and `redirect_uri` values from our application configuration. 
+
+We will also construct the URL used to make a [Logout](https://fusionauth.io/docs/v1/tech/oauth/endpoints/#logout) request.
+
+```ruby
+# app/controllers/welcome_controller.rb
+
+class WelcomeController < ApplicationController
+  def initialize
+    @login_url = 'http://localhost:9011/oauth2/authorize?client_id=799b3fe1-4a52-494f-a13a-b97b1377e073&response_type=code&redirect_uri=http%3A%2F%2Flocalhost%3A3000%2Foauth2-callback'
+    @logout_url = 'http://localhost:9011/oauth2/logout?client_id=799b3fe1-4a52-494f-a13a-b97b1377e073'
+  end
+end
+```
+
+
+I decided that I wanted to utilize my environment to set my `client_id`, `client_secret` and `idp_url` (identity provider url) using `development.rb`.
+```ruby
+  # OAuth configuration
+  config.x.oauth.client_id = "my-client-id"
+  config.x.oauth.client_secret = "my-super-secret-oauth-secret"
+  config.x.oauth.idp_url = "http://localhost:9011/"
+```
+
+Additionally, I added the following to `ApplicationController` such that my child classes have access.
+```ruby
+  protected
+
+  def idp_url
+    Rails.configuration.x.oauth.idp_url
+  end
+
+  def client_id
+    Rails.configuration.x.oauth.client_id
+  end
+
+  def client_secret
+    Rails.configuration.x.oauth.client_secret
+  end
+```
+
+Now it's time to actually implement the OAuth callback. 
+
+```ruby
+# app/controllers/oauth_controller.rb
+
+class OauthController < ApplicationController
+  def initialize
+    @redirect_uri = "http://localhost:3000/oauth2-callback"
```

Comment:
Since this is used in a couple of places, should it be extracted to the config file?